charmbits/service/server.go

```diff
@@ -49,29 +46,56 @@ type Context struct {
 	socketPath string
 }
 
+type rpcCommand struct {
+	tomb     tomb.Tomb
+	listener net.Listener
+}
+
+func (c *rpcCommand) Kill() {
+	c.listener.Close()
+}
+
+func (c *rpcCommand) Wait() error {
+	return c.tomb.Wait()
+}
+
+func (c *rpcCommand) run(srv *rpc.Server) {
+	for {
+		conn, err := c.listener.Accept()
+		if err != nil {
+			log.Printf("local socket accept failed: %v", err)
+			return
+		}
+		log.Printf("local RPC accepted dial request")
+		// TODO shut down existing RPC requests cleanly
+		go srv.ServeCodec(jsonrpc.NewServerCodec(conn))
+	}
+}
+
 // ServeLocalRPC starts a local RPC server serving methods on the given
 // receiver value, using the net/rpc package (see rpc.Server.Register).
 //
 // The methods may be invoked using the Service.Call method. Parameters
 // and return values will be marshaled as JSON.
 //
-// ServeLocalRPC blocks indefinitely.
-func (ctxt *Context) ServeLocalRPC(rcvr interface{}) error {
+// ServeLocalRPC returns the Command representing the running
+// service.
+func (ctxt *Context) ServeLocalRPC(rcvr interface{}) (hook.Command, error) {
 	srv := rpc.NewServer()
 	srv.Register(rcvr)
 	listener, err := listen(ctxt.socketPath)
 	if err != nil {
-		return errgo.Notef(err, "cannot listen on local socket")
+		return nil, errgo.Notef(err, "cannot listen on local socket")
 	}
-	for {
-		log.Printf("accepting local service on %s", ctxt.socketPath)
-		conn, err := listener.Accept()
-		if err != nil {
-			return errgo.Notef(err, "local socket accept failed")
-		}
-		log.Printf("accepted dial request")
-		go srv.ServeCodec(jsonrpc.NewServerCodec(conn))
+	cmd := &rpcCommand{
+		listener: listener,
 	}
+	log.Printf("accepting local service on %s", ctxt.socketPath)
+	cmd.tomb.Go(func() error {
+		cmd.run(srv)
+		return nil
+	})
```

Comment:
I don't *think* so as there's only one goroutine, so if that
goroutine exits, the tomb will immediately be done,
and the only way to put the tomb into dying state is to
kill it, which closes the listener which should cause the goroutine
to exit anyway.

We could take the listener.Close out of the Kill method
and make the Kill method kill the tomb instead and
have your suggested goroutine in place. Do you think
that would be better?